From 10602ca53b1c1231e96d8d446904617528c11dae Mon Sep 17 00:00:00 2001
From: Ronan Le Martret <ronan.lemartret@iot.bzh>
Date: Mon, 13 Mar 2017 15:31:41 +0100
Subject: [PATCH] Fix missing link to unistring for lms plugin

Signed-off-by: Ronan Le Martret <ronan.lemartret@iot.bzh>
---
 configure.ac | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/configure.ac b/configure.ac
index c5cbc00..2f7dc78 100644
--- a/configure.ac
+++ b/configure.ac
@@ -248,6 +248,10 @@ AS_IF([test "x$enable_lms_plugin" = "xyes"],
                        gio-2.0 >= $GIO_REQUIRED
                        sqlite3 >= $LIBSQLITE3_REQUIRED])
     RYGEL_PLUGIN_LMS_DEPS_VALAFLAGS="$RYGEL_COMMON_MODULES_VALAFLAGS --pkg gio-2.0 --pkg gee-0.8 --pkg sqlite3"
+    AS_IF([test "x$have_unistring" = "xyes"],
+      [
+        RYGEL_PLUGIN_LMS_DEPS_LIBS="$RYGEL_PLUGIN_LMS_DEPS_LIBS -lunistring"
+      ])
     AC_SUBST([RYGEL_PLUGIN_LMS_DEPS_VALAFLAGS])
   ])
 
-- 
2.6.6

