From 26927f2aff43607b422ce5cb463ee4eccbdaef5a Mon Sep 17 00:00:00 2001
From: Ronan Le Martret <ronan.lemartret@iot.bzh>
Date: Thu, 23 Feb 2017 14:18:06 +0100
Subject: [PATCH] Fix arm-trusted-firmware build for gcc6

Signed-off-by: Ronan Le Martret <ronan.lemartret@iot.bzh>
---
 include/lib/smcc.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/lib/smcc.h b/include/lib/smcc.h
index c415ba1..e648692 100644
--- a/include/lib/smcc.h
+++ b/include/lib/smcc.h
@@ -57,7 +57,7 @@
 #define SMC_64				1
 #define SMC_32				0
 #define SMC_UNK				0xffffffff
-#define SMC_TYPE_FAST			1
+#define SMC_TYPE_FAST			1U
 #define SMC_TYPE_STD			0
 #define SMC_PREEMPTED		0xfffffffe
 /*******************************************************************************
-- 
2.6.6

