--- a/Makefile.am
+++ b/Makefile.am
@@ -10,5 +10,5 @@ ACLOCAL_AMFLAGS = -I autoconf
 
 #@SET_MAKE@
 
-SUBDIRS = lib libtar doc
+SUBDIRS = lib libtar doc testsuite
 
--- a/doc/Makefile.in
+++ b/doc/Makefile.in
@@ -151,3 +151,4 @@ install: all
 		echo ".so man3/@LISTHASH_PREFIX@_list_new.3" > ${DESTDIR}${mandir}/man3/$${i}.3; \
 	done
 
+check:
--- a/lib/Makefile.in
+++ b/lib/Makefile.in
@@ -104,3 +104,4 @@ install: ${ALL}
 	${INSTALL_DATA} ${srcdir}/libtar.h ${DESTDIR}${includedir}
 	${INSTALL_DATA} ../listhash/libtar_listhash.h ${DESTDIR}${includedir}
 
+check:
--- a/libtar/Makefile.in
+++ b/libtar/Makefile.in
@@ -76,3 +76,4 @@ install: ${ALL}
 	${MKDIR} ${DESTDIR}${bindir}
 	$(LIBTOOL) --mode=install ${INSTALL_PROGRAM} libtar ${DESTDIR}${bindir}
 
+check:
--- /dev/null
+++ b/testsuite/Makefile.in
@@ -0,0 +1,7 @@
+all:
+
+check: ../libtar/libtar
+	../libtar/libtar -C ../doc -c test.tar .
+	../libtar/libtar -t test.tar
+	$(RM) test.tar
+install:
--- a/configure.ac
+++ b/configure.ac
@@ -120,6 +120,6 @@ fi
 
 
 dnl ### Create output files. #######################################
-AC_CONFIG_FILES([Makefile lib/Makefile libtar/Makefile doc/Makefile])
+AC_CONFIG_FILES([Makefile lib/Makefile libtar/Makefile doc/Makefile testsuite/Makefile])
 AC_OUTPUT
 
